### DISCUSSION TITLE
extended domain为什么这么计算？

### DISCUSSION THREAD

**Discussion by mdj33** [z2o-k7e/halo2-step-by-step#29] [Q&A] [answered]
在 计算domain时候会计算extended domain,  代码注释里说为了计算商多项式， 从日志看，一般是扩大四倍，也就是i是2，
比如
degree=3, k=4, extend_k=6
degree=3, k=5, extend_k=7
请问，为什么是这样计算？ 并且为什么扩大四倍，就够了呢？谢谢

// quotient_poly_degree * params.n - 1 is the degree of the quotient polynomial
        let quotient_poly_degree = (j - 1) as u64;

        // n = 2^k
        let n = 1u64 << k;

        // We need to work within an extended domain, not params.k but params.k + i
        // for some integer i such that 2^(params.k + i) is sufficiently large to
        // describe the quotient polynomial.
        let mut extended_k = k;
        while (1 << extended_k) < (n * quotient_poly_degree) {
            extended_k += 1;
        }

代码地址: https://github.com/zcash/halo2/blob/f9838c127ec9c14f6f323e0cfdc0c1392594d37f/halo2_proofs/src/poly/domain.rs#L47

**Comment by Demian101** [comment]
在 Plonkathon（PLONK 协议的一个实现） 的代码中：
https://github.com/0xPARC/plonkathon/blob/main/poly.py#L150-L177

```python
    # Converts a list of evaluations at [1, w, w**2... w**(n-1)] to
    # a list of evaluations at
    # [offset, offset * q, offset * q**2 ... offset * q**(4n-1)] where q = w**(1/4)
    # This lets us work with higher-degree polynomials, and the offset lets us
    # avoid the 0/0 problem when computing a division (as long as the offset is
    # chosen randomly)
    def to_coset_extended_lagrange(self, offset):
        assert self.basis == Basis.LAGRANGE
        group_order = len(self.values)
        x_powers = self.ifft().values
        x_powers = [(offset**i * x) for i, x in enumerate(x_powers)] + [Scalar(0)] * (
            group_order * 3
        )
        return Polynomial(x_powers, Basis.MONOMIAL).fft()

    # Convert from offset form into coefficients
    # Note that we can't make a full inverse function of to_coset_extended_lagrange
    # because the output of this might be a deg >= n polynomial, which cannot
    # be expressed via evaluations at n roots of unity
    def coset_extended_lagrange_to_coeffs(self, offset):
        assert self.basis == Basis.LAGRANGE

        shifted_coeffs = self.ifft().values
        inv_offset = 1 / offset
        return Polynomial(
            [v * inv_offset**i for (i, v) in enumerate(shifted_coeffs)],
            Basis.MONOMIAL,
        )
```

将一系列在 [1, w, w**2... w**(n-1)] 的 evaluations 转换为在 [offset, offset * q, offset * q**2 ... offset * q**(4n-1)] 

细节可能需要老师们帮忙确认下 ~

**Reply by mdj33** [reply]
done

**Comment by Demian101** [comment]
> ```
>    在 计算domain时候会计算extended domain,  代码注释里说为了计算商多项式， 从日志看，一般是扩大四倍，也就是i是2，
> ```
> 
> 比如 degree=3, k=4, extend_k=6 degree=3, k=5, extend_k=7 请问，为什么是这样计算？ 并且为什么扩大四倍，就够了呢？谢谢
> 
> // quotient_poly_degree * params.n - 1 is the degree of the quotient polynomial let quotient_poly_degree = (j - 1) as u64;
> 
> ```
>     // n = 2^k
>     let n = 1u64 << k;
> 
>     // We need to work within an extended domain, not params.k but params.k + i
>     // for some integer i such that 2^(params.k + i) is sufficiently large to
>     // describe the quotient polynomial.
>     let mut extended_k = k;
>     while (1 << extended_k) < (n * quotient_poly_degree) {
>         extended_k += 1;
>     }
> ```

具体的代码地址可以贴一下嘛

**Reply by CeciliaZ030** [reply, marked as answer]
扩大4倍的说法错的。就是比如你 k=8 就是电路有256行，每行的expression最高deg=3，那我就需要3*256 = 768的总degree
那么多大的k能cover 768？明显是1024，即k=10。那个while loop就算这个。

**Comment by Demian101** [comment]
💧ma: 明白了， 就是为了把这个degree的3 cover进去的2的最小pow，字面上看也就是2^2

这个extended k 的pow也是里面ExtendedLagrangeCoeff的最大阶吧？

> 对
